datastore/datastore.go

```diff
@@ -31,185 +26,190 @@ var (
+	PersistentVolumeInformer      cache.SharedInformer
+	pvcLister                     corelisters.PersistentVolumeClaimLister
+	PersistentVolumeClaimInformer cache.SharedInformer
+	cfmLister                     corelisters.ConfigMapLister
+	ConfigMapInformer             cache.SharedInformer
+	secretLister                  corelisters.SecretLister
+	SecretInformer                cache.SharedInformer
+	knLister                      corelisters.NodeLister
+	KubeNodeInformer              cache.SharedInformer
+	pcLister                      schedulinglisters.PriorityClassLister
+	PriorityClassInformer         cache.SharedInformer
+	csiDriverLister               storagelisters_v1.CSIDriverLister
+	CSIDriverInformer             cache.SharedInformer
+	storageclassLister            storagelisters_v1.StorageClassLister
+	StorageClassInformer          cache.SharedInformer
+	pdbLister                     policylisters.PodDisruptionBudgetLister
+	PodDistrptionBudgetInformer   cache.SharedInformer
+	svLister                      corelisters.ServiceLister
+	ServiceInformer               cache.SharedInformer
 }
 
 // NewDataStore creates new DataStore object
 func NewDataStore(
-	volumeInformer lhinformers.VolumeInformer,
-	engineInformer lhinformers.EngineInformer,
-	replicaInformer lhinformers.ReplicaInformer,
-	engineImageInformer lhinformers.EngineImageInformer,
-	nodeInformer lhinformers.NodeInformer,
-	settingInformer lhinformers.SettingInformer,
-	imInformer lhinformers.InstanceManagerInformer,
-	smInformer lhinformers.ShareManagerInformer,
-	biInformer lhinformers.BackingImageInformer,
-	bimInformer lhinformers.BackingImageManagerInformer,
-	bidsInformer lhinformers.BackingImageDataSourceInformer,
-	btInformer lhinformers.BackupTargetInformer,
-	bvInformer lhinformers.BackupVolumeInformer,
-	bInformer lhinformers.BackupInformer,
-	rjInformer lhinformers.RecurringJobInformer,
+	lhInformerFactory lhinformers.SharedInformerFactory,
 	lhClient lhclientset.Interface,
-
-	podInformer coreinformers.PodInformer,
-	cronJobInformer batchinformers_v1beta1.CronJobInformer,
-	daemonSetInformer appsinformers.DaemonSetInformer,
-	deploymentInformer appsinformers.DeploymentInformer,
-	persistentVolumeInformer coreinformers.PersistentVolumeInformer,
-	persistentVolumeClaimInformer coreinformers.PersistentVolumeClaimInformer,
-	configMapInformer coreinformers.ConfigMapInformer,
-	secretInformer coreinformers.SecretInformer,
-	kubeNodeInformer coreinformers.NodeInformer,
-	priorityClassInformer schedulinginformers.PriorityClassInformer,
-	csiDriverInformer storageinformers_v1.CSIDriverInformer,
-	storageclassInformer storageinformers_v1.StorageClassInformer,
-	pdbInformer policyinformers.PodDisruptionBudgetInformer,
-	serviceInformer coreinformers.ServiceInformer,
-
+	kubeInformerFactory informers.SharedInformerFactory,
 	kubeClient clientset.Interface,
 	namespace string) *DataStore {
 
+	replicaInformer := lhInformerFactory.Longhorn().V1beta1().Replicas()
```

Comment:
NIT: let's do the same as in the controllers, (informer, cache) pair.
Makes it easy to see that a cachesync add to the array is necessary when adding a new informer :)

Reply:
Sure you add `cacheSyncs []cache.InformerSynced` to the datastore struct.
Then after each informer, you append to the cacheSyncs array, just like in the controllers.
Then in the DS.sync function you wait for all hasSynced functions, just like in the controllers.

```go
func (s *DataStore) Sync(stopCh <-chan struct{}) bool {
	return cache.WaitForNamedCacheSync("longhorn datastore", stopCh, cacheSyncs...)
}
```